商品購入機能 #10

Tomogit2 opened the pull request:
#Why
商品購入機能を実装するため。

#What
商品をクレジットカード決済で購入する機能を実装から単体テストまで行った。


・必要な情報を適切に入力して「購入」ボタンを押すと、商品の購入ができる動画
https://gyazo.com/7ffc6284d2adf451679c044b56bbd5ab

・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/fcf88fda58e20fd902108b1cb1360236

・ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移する動画
https://gyazo.com/23e09e96436188066c13871790e1f544

・ログイン状態の場合でも、URLを直接入力して自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移する動画
https://gyazo.com/d4724e1397c9e7c061e036d68f8086d1

・ログアウト状態の場合は、URLを直接入力して商品購入ページに遷移しようとすると、商品の販売状況に関わらずログインページに遷移する動画
https://gyazo.com/5d6b74f7e04549def5012b9b0ef9e7d7

・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品一覧機能実装時に未実装であった場合）
・売却済みの商品は、画像上に「sold out」の文字が表示される動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/08418bd4a30ac0a644c2b5b7018d4adf

・ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されない動画（商品詳細機能実装時に未実装であった場合）
https://gyazo.com/4d2bb72bd900d93d48600d6178acb5bc

・ログイン状態の場合でも、URLを直接入力して自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移する動画（商品情報編集機能実装時に未実装であった場合）
https://gyazo.com/cda527ad6c5db6582d91483803891314

・テスト結果の画像
https://gyazo.com/ef994cd9addb56ce9d1979e08d8d34a6
https://gyazo.com/b8674e296828468f9ce8de298947c7bd

Tomogit2 committed:
PAY.JPを使用するために、Turboの機能をオフにした。

事前準備 不具合を回避するためTurboをオフにした。

Tomogit2 committed:
purchaseモデルを作成した。

マイグレーションファイルを変更した。purchasesモデルにアソシエーションとバリデーションを追記した。

Tomogit2 committed:
newとcreateアクションのルーティングを記述した。

Tomogit2 committed:
新たにFormオブジェクトを作成した。

新たにmodelsディレクトリ直下にファイルを作成し、クラスを定義した。

Tomogit2 committed:
Formオブジェクトを記述した。

作成したクラスでActiveModel::Modelを呼び出して、form_withメソッドに対応する機能とバリデーションを行う機能を持たせた。
attr_accessorを用いて、保存したい複数のテーブルのカラム名すべてを扱えるようにした。
バリデーションの処理を書いた。
データをテーブルに保存する処理を書いた。

Tomogit2 committed:
コントローラーのnewアクション、createアクションでFormオブジェクトのインスタンスを生成するようにした。

Tomogit2 committed:
ルーティングの設定をした。

Tomogit2 committed:
Purchasesコントローラーの修正

Tomogit2 committed:
purchaseのビューファイルを作成した。

Tomogit2 committed:
購入のためのフォームを実装した。

Tomogit2 committed:
ログイン状態の場合でも、自身が出品した商品の商品購入ページに遷移しようとすると、商品の販売状況に関わらずトップページに遷移するようにした。

Tomogit2 committed:
PAY.JPを用いたクレジットカードでの決済機能の実装

Tomogit2 committed:
売却済みの商品は、画像上に「sold out」の文字が表示されるようにした。

商品一覧機能、商品詳細機能。

Tomogit2 committed:
ログイン状態の場合でも、売却済みの商品には、「商品の編集」「削除」「購入画面に進む」ボタンが表示されないようにした。

（商品詳細機能）

Tomogit2 committed:
ログイン状態の場合でも、自身が出品した売却済み商品の商品情報編集ページへ遷移しようとすると、トップページに遷移するようにした。

（商品情報編集機能）

Tomogit2 committed:
モデルの単体テストコードを記述した。

Tomogit2 committed:
Rubocopを実行し、check内容を自動修正した。

Tomogit2 committed:
ログイン状態の場合でも、URLを直接入力して自身が出品していない売却済み商品の商品購入ページへ遷移しようとすると、トップページに遷移するようにした。

Tomogit2 committed:
必要な情報を適切に入力して「購入」ボタンを押すと、DBに購入情報が保存されるようにした。

Tomogit2 committed:
PurchaseAddressモデルのテストコードを記述した。

Tomogit2 committed:
ペンディングの通知を引き起こしていたファイルを削除した。

minachanchan reviewed (commented):
レビューのご依頼ありがとうございます！
キャプチャのご提示もお礼申し上げます。

以下挙動について、見本アプリと同様の実装になるよう見直しましょう！

```
・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/fcf88fda58e20fd902108b1cb1360236

エラーメッセージの出力場所が見本アプリと同じになるように実装を確認しましょう。
```
実装、挙動をご確認の上、再度Gyazoで撮影いただいたキャプチャのURLをこちらのプルリクエストに添付いただければと存じます。
その後、再度レビュー依頼いただけますと幸いです。

何卒よろしくお願いいたします。

Tomogit2 committed:
入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示されるようにした。

Tomogit2 commented:
ご返信ありがとうございます。

・入力に問題がある状態で「購入」ボタンが押された場合、情報は受け入れられず、購入ページでエラーメッセージが表示される動画
https://gyazo.com/4632965916e9a329fd92a3a669353eae

です。よろしくお願いいたします。

koyamatakumi commented on app/controllers/purchases_controller.rb:
```diff
@@ -0,0 +1,63 @@
+      Rails.logger.error @purchase_address.errors.full_messages.join(", ")
+      gon.public_key = ENV["PAYJP_PUBLIC_KEY"]
+      render 'new', status: :unprocessable_entity
+  end
```
```ruby
    if @purchase_address.valid?
      pay_item
      if @purchase_address.save
        redirect_to root_path
      else
        Rails.logger.error @purchase_address.errors.full_messages.join(", ")
        gon.public_key = ENV["PAYJP_PUBLIC_KEY"]
        render 'new', status: :unprocessable_entity
      end
    else
      Rails.logger.error @purchase_address.errors.full_messages.join(", ")
      gon.public_key = ENV["PAYJP_PUBLIC_KEY"]
      render 'new', status: :unprocessable_entity
  end
```
上記の部分のリファクタリングを行いましょう。
■補足
@purchase_address.valid?でバリデーションのチェックは既に行われているため、「 if @purchase_address.save」を条件式にする必要はありません。

Tomogit2 replied:
修正いたしました。

koyamatakumi commented on app/controllers/purchases_controller.rb:
```diff
@@ -0,0 +1,63 @@
+  end
+
+  def create
+    @purchase_address = PurchaseAddress.new(purchase_address_params.merge(user_id: current_user.id, item_id: params[:item_id]))
```
「.merge(user_id: current_user.id, item_id: params[:item_id])」はストロングパラメーターに記載しましょう。

Tomogit2 replied:
ストロングパラメーターに記載しましたが、少し不安です。ご確認よろしくお願いいたします。

koyamatakumi commented on app/javascript/application.js:
```diff
@@ -2,3 +2,5 @@
 import "controllers"
 import "sales_fee"
+import "card"
+console.log(window.payjpPublicKey);
```
```js
console.log(window.payjpPublicKey);
```
デバッグで使用した記述は消しておきましょう。

Tomogit2 replied:
ありがとうございます。削除いたしました。

koyamatakumi commented on app/javascript/card.js:
```diff
@@ -0,0 +1,103 @@
+// document.addEventListener("DOMContentLoaded", function() {
```
不要なコメントアウトは消しましょう。

Tomogit2 replied:
失礼いたしました。削除いたしました。

koyamatakumi commented on app/models/purchase_address.rb:
```diff
@@ -0,0 +1,32 @@
+                           street_address:, building_name:, telephone_number:, purchase_id: purchase.id)
+    true
+=end
+  end
```
```ruby
  def save
    return false unless valid?
    purchase = Purchase.create(item_id: item_id, user_id: user_id)
    if purchase.persisted?
      ShippingAddress.create(zip_code: zip_code, prefecture_id: prefecture_id, municipalities: municipalities,
                             street_address: street_address, building_name: building_name, telephone_number: telephone_number, purchase_id: purchase.id)
      true
    else
      Rails.logger.error "Failed to create Purchase: #{purchase.errors.full_messages.join(", ")}"
      false
    end

=begin
    purchase = Purchase.create(item_id:, user_id:)
    ShippingAddress.create(zip_code:, prefecture_id:, municipalities:,
                           street_address:, building_name:, telephone_number:, purchase_id: purchase.id)
    true
=end
  end
```
リファクタリングを行いましょう。
この時点ではバリデーションのチェックは不要です。（コントローラーでチェックしているため）
各テーブルへ保存する処理のみが書かれていれば問題ありません。Formオブジェクトの補足カリキュラムを参考に修正してみましょう。

Tomogit2 replied:
修正いたしました。

koyamatakumi commented on app/models/shipping_address.rb:
```diff
@@ -0,0 +1,9 @@
+  validates :prefecture_id, presence: true
+  validates :municipalities, presence: true
+  validates :street_address, presence: true
+  validates :telephone_number, presence: true
```
```ruby
  validates :zip_code, presence: true
  validates :prefecture_id, presence: true
  validates :municipalities, presence: true
  validates :street_address, presence: true
  validates :telephone_number, presence: true
```
Formオブジェクト内にバリデーションを記載しているため不要となります。

Tomogit2 replied:
修正いたしました。

koyamatakumi commented on app/views/items/index.html.erb:
```diff
@@ -133,20 +132,16 @@
+                <%= image_tag item.image, class: "item-img" %>
+              <% else %>
+                <%= image_tag "item-sample.png", class: "item-img" %>
+              <% end %>
```
```ruby
              <% if item.image.attached? %>
                <%= image_tag item.image, class: "item-img" %>
              <% else %>
                <%= image_tag "item-sample.png", class: "item-img" %>
              <% end %>
```
出品時に商品の画像は必須にしているため、条件式は不要となります。
「 <%= image_tag item.image, class: "item-img" %>」以外消しておきましょう。

Tomogit2 replied:
修正いたしました。

koyamatakumi commented on app/views/items/show.html.erb:
```diff
@@ -23,14 +25,12 @@
+      <%= link_to "削除", item_path(@item), data: {turbo_method: :delete}, class:"item-destroy" %>
+    <% elsif user_signed_in? && current_user != @item.user && @item.purchase.blank? %>
+      <%= link_to "購入画面に進む", new_item_purchase_path(@item), data: { turbo: false },class:"item-red-btn"%>
     <% end %>
```
```ruby
    <% if current_user == @item.user && @item.purchase.blank? %>
      <%= link_to "商品の編集", edit_item_path(@item), method: :get, class: "item-red-btn" %>
      <p class="or-text">or</p>
      <%= link_to "削除", item_path(@item), data: {turbo_method: :delete}, class:"item-destroy" %>
    <% elsif user_signed_in? && current_user != @item.user && @item.purchase.blank? %>
      <%= link_to "購入画面に進む", new_item_purchase_path(@item), data: { turbo: false },class:"item-red-btn"%>
    <% end %>
```
リファクタリングの余地があるため修正しましょう。
以下のように記述することができます。
```
<if ログインしていたら　且つ　商品が売れていなければ>
　　　　<if　ログインユーザーと出品者が同じならば>
　　　　<編集>
　　　　<削除>
　　　　<else>
　　　　<購入画面に進む>
　　　　<end>
<end>
```

Tomogit2 replied:
ありがとうございます。修正いたしました。

koyamatakumi commented on spec/factories/purchase_addresses.rb:
```diff
@@ -0,0 +1,14 @@
+FactoryBot.define do
+  factory :purchase_address do
+    association :user
+    association :item
```
```rubyt
    association :user
    association :item
```
Formオブジェクトはモデルではないためアソシエーションの記述は使用できません。
こちらは消しておきましょう。

Tomogit2 replied:
修正いたしました。

koyamatakumi commented on spec/forms/purchase_address_spec.rb:
```diff
@@ -0,0 +1,64 @@
+RSpec.describe PurchaseAddress, type: :model do
+  before do
+    @user = FactoryBot.create(:user)
+    @item = FactoryBot.create(:item, user: @user)
```
```ruby
    @item = FactoryBot.create(:item, user: @user)
```
「, user: @user」は不要です。消しておきましょう。
■理由
itemのFactoryBotには「   association :user」が記載されているため、商品インスタンスを作成時にユーザーを紐づけることができている。また、現状では、出品者と購入者が同一のユーザーになってしまっているため。

Tomogit2 replied:
分かりやすいご説明ありがとうございます。修正いたしました。

koyamatakumi commented on spec/forms/purchase_address_spec.rb:
```diff
@@ -0,0 +1,64 @@
+      it '全ての値が正しく入力されていれば購入できる' do
+        expect(@purchase_address).to be_valid
+      end
+    end
```
```ruby
    context '商品購入ができるとき' do
      it '全ての値が正しく入力されていれば（tokenがあれば）購入できる' do
        expect(@purchase_address).to be_valid
      end
    end

    context '商品購入ができないとき' do
      it 'tokenが空では登録できないこと' do
        @purchase_address.token = nil
        @purchase_address.valid?
        expect(@purchase_address.errors.full_messages).to include("Token can't be blank")
      end
    end

    context '商品購入ができるとき' do
      it '全ての値が正しく入力されていれば購入できる' do
        expect(@purchase_address).to be_valid
      end
    end
```
contextが重複している為以下のようにまとめましょう。
```
context '商品購入ができるとき' do
正常系テストを囲む
end
context '商品購入ができないとき' do
異常系テストを囲む
end

```

Tomogit2 replied:
修正いたしました。

koyamatakumi commented on spec/forms/purchase_address_spec.rb:
```diff
@@ -0,0 +1,64 @@
+        @purchase_address.telephone_number = nil
+        @purchase_address.valid?
+        expect(@purchase_address.errors.full_messages).to include("Telephone number can't be blank")
+      end
```
異常系のテストが不足しているため、バリデーションを確認しながら追加しましょう。

Tomogit2 replied:
テストを追記しました。
・テスト結果の画像
https://gyazo.com/f434a42dd2e5aa9dfd3de20442bd1a8a
https://gyazo.com/4a9db21931063f3be68a91ee6b7eab58
https://gyazo.com/57b28f679be216cc1251f99b3596b128

koyamatakumi reviewed (commented):
Tomogit2様、レビュー依頼フォームの送信ありがとうございます。
コードを確認させていただき、コメントいたしましたのでご確認をお願いいたします。
修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します！

Tomogit2 committed:
if @purchase_address.valid?のリファクタリング

@purchase_address.valid?でバリデーションのチェックは既に行われているため、「 if @purchase_address.save」を条件式にする必要がない。

Tomogit2 committed:
「.merge(user_id: current_user.id, item_id: params[:item_id])」をストロングパラメーターに記載した。

Tomogit2 committed:
不要な記述の削除

Tomogit2 committed:
リファクタリングを行った。

バリデーションのチェックは不要。（コントローラーでチェックしているため）
各テーブルへ保存する処理のみが書かれていれば問題ない。

Tomogit2 committed:
Formオブジェクト内にバリデーションを記載しているため削除した。

Tomogit2 committed:
出品時に商品の画像は必須にしているため、条件式を削除した。

Tomogit2 committed:
リファクタリングした。

Tomogit2 committed:
アソシエーションの記述を削除した。

Formオブジェクトはモデルではないためアソシエーションの記述は使用できないため。

Tomogit2 committed:
「, user: @user」を削除した。

itemのFactoryBotには「 association :user」が記載されているため、商品インスタンスを作成時にユーザーを紐づけることができている。また、現状では、出品者と購入者が同一のユーザーになってしまっているため。

Tomogit2 committed:
Update purchase_address_spec.rb

重複しているcontextをまとめた。

Tomogit2 committed:
テストを追記した。

koyamatakumi commented on app/controllers/purchases_controller.rb:
```diff
@@ -0,0 +1,60 @@
+
+  def set_payjp_api_key
+    Payjp.api_key = ENV['PAYJP_SECRET_KEY']
+  end
```
```ruby
  def set_payjp_api_key
    Payjp.api_key = ENV['PAYJP_SECRET_KEY']
  end
```
private以下に書きましょう。

Tomogit2 replied:
修正いたしました。

koyamatakumi commented on app/models/purchase_address.rb:
```diff
@@ -0,0 +1,22 @@
+    else
+      Rails.logger.error "Failed to create Purchase: #{purchase.errors.full_messages.join(", ")}"
+    end
+  end
```
```ruby
  def save
    purchase = Purchase.create(item_id: item_id, user_id: user_id)
    if purchase.persisted?
      ShippingAddress.create(zip_code: zip_code, prefecture_id: prefecture_id, municipalities: municipalities,
                             street_address: street_address, building_name: building_name, telephone_number: telephone_number, purchase_id: purchase.id)
    else
      Rails.logger.error "Failed to create Purchase: #{purchase.errors.full_messages.join(", ")}"
    end
  end
```
以下のようにシンプルな記述で問題ないかと存じます。
```ruby
  def save
    purchase = Purchase.create(item_id: item_id, user_id: user_id)
      ShippingAddress.create(zip_code: zip_code, prefecture_id: prefecture_id, municipalities: municipalities,
                             street_address: street_address, building_name: building_name, telephone_number: telephone_number, purchase_id: purchase.id)
  end
```

Tomogit2 replied:
修正いたしました。

koyamatakumi commented on spec/forms/purchase_address_spec.rb:
```diff
@@ -0,0 +1,106 @@
+    context '商品購入ができるとき' do
+      it '全ての値が正しく入力されていれば（tokenがあれば）購入できる' do
+        expect(@purchase_address).to be_valid
+      end
```
「建物名が空でも保存できる」という正常系テストを追加しましょう。

Tomogit2 replied:
追加いたしました。

Tomogit2 replied:
・テスト結果の画像
https://gyazo.com/06527597a2c6ed6e45cd60562eb895b7
https://gyazo.com/7458890af8550d3dd0a2522207f48249

koyamatakumi reviewed (commented):
Tomogit2様
レビュー依頼フォームの送信ありがとうございます。 修正いただいた内容を確認させていただき、追加でコメントいたしましたのでご確認お願いいたします。
修正後はコミット、プッシュを忘れずに行い、改めてレビュー依頼フォームの送信をお願い致します。着実にゴールに近づいています。ラストスパート頑張っていきましょう！

Tomogit2 committed:
private以下に移動した。

Tomogit2 committed:
シンプルな記述にした。

Tomogit2 committed:
「建物名が空でも保存できる」という正常系テストを追加した。

koyamatakumi commented:
Tomogit2様
レビュー依頼フォームの送信ありがとうございます。
確認したところ、**前回のレビュー以降の差分が見つかりませんでした。**
コミットプッシュのし忘れがないかを確認いただき、差分を上げていただいたら改めてレビュー依頼フォームの送信をお願いいたします。

Tomogit2 commented:
大変失礼いたしました。コミットプッシュのし忘れでした。お手数をおかけしますがよろしくお願いいたします。

minachanchan reviewed (commented):
レビューのご依頼ありがとうございます。

実装内容を確認させていただきました。
**LGTMとさせていただきます。実装お疲れ様でした！**

このブランチをマージし、masterブランチ（もしくはmainブランチ）でプルしましょう。
内容が反映されていることを確認し、次の実装・学習に入りましょう！
引き続き応援しております！
